cpp/velox/memory/VeloxMemoryPool.cc

```diff
@@ -122,91 +95,126 @@ class VeloxMemoryAllocator final : public velox::memory::MemoryAllocator {
+    pool->grow(bytes);
+  }
+
+  uint64_t releaseMemory(velox::memory::MemoryPool* pool, uint64_t bytes) override {
+    uint64_t freeBytes = pool->shrink(bytes);
+    listener_->allocationChanged(-freeBytes);
+    if (bytes == 0 && pool->capacity() != 0) {
+      // So far only MemoryManager::dropPool() calls with 0 bytes. Let's assert the pool
+      //   gives all capacity back to Spark
+      //
+      // We are likely in destructor, do not throw. INFO log is fine since we have leak checks from Spark's memory
+      //   manager
+      LOG(INFO) << "Memory pool " << pool->name() << " not completely shrunk when Memory::dropPool() is called";
     }
-    return child;
+    return freeBytes;
+  }
+
+  bool growMemory(
+      velox::memory::MemoryPool* pool,
+      const std::vector<std::shared_ptr<velox::memory::MemoryPool>>& candidatePools,
+      uint64_t targetBytes) override {
+    GLUTEN_CHECK(candidatePools.size() == 1, "ListenableArbitrator should only be used within a single root pool");
+    auto candidate = candidatePools.back();
+    GLUTEN_CHECK(pool->root() == candidate.get(), "Illegal state in ListenableArbitrator");
+    reserveMemory(pool, targetBytes);
+    return true;
+  }
+
+  Stats stats() const override {
+    Stats stats; // no-op
+    return stats;
+  }
+
+  std::string toString() const override {
+    return fmt::format(
+        "ARBITRATOR[{}] CAPACITY {} {}", kindString(kind_), velox::succinctBytes(capacity_), stats().toString());
   }
 
  private:
-  std::shared_ptr<velox::memory::MemoryAllocator> sharedAlloc_; // workaround to keep the allocator instance alive
-};
+  void abort(velox::memory::MemoryPool* pool) {
+    try {
+      pool->abort();
+    } catch (const std::exception& e) {
+      LOG(WARNING) << "Failed to abort memory pool " << pool->toString();
+    }
+    // NOTE: no matter memory pool abort throws or not, it should have been marked
+    // as aborted to prevent any new memory arbitration triggered from the aborted
+    // memory pool.
+    VELOX_CHECK(pool->aborted());
+  }
 
-static std::shared_ptr<velox::memory::MemoryPool> rootVeloxMemoryPool() {
-  static auto options = gluten::VeloxInitializer::get()->getMemoryPoolOptions();
-  static int64_t spillThreshold = gluten::VeloxInitializer::get()->getSpillThreshold();
-  static std::shared_ptr<VeloxMemoryPool> defaultPoolRoot = std::make_shared<VeloxMemoryPool>(
-      nullptr,
-      "root",
-      velox::memory::MemoryPool::Kind::kAggregate,
-      getDefaultVeloxMemoryManager(),
-      facebook::velox::memory::MemoryReclaimer::create(),
-      nullptr,
-      options);
-
-  auto f = [&]() {
-    defaultPoolRoot->setHighUsageCallback(
-        [=](velox::memory::MemoryPool& pool) { return pool.reservedBytes() >= spillThreshold; });
-    defaultPoolRoot->testingSetCapacity(facebook::velox::memory::kMaxMemory);
-    return true;
-  };
+  gluten::AllocationListener* listener_;
+};
 
-  static bool do_once = f();
-  (void)do_once; // suppress warning
```

Comment:
why not use `MemoryManager::getInstance()` instead

Reply:
I'm OK to either using this approach or `MemoryManager::getInstance()`.